index.js

```diff
@@ -64,18 +65,27 @@ class MergedStream extends PassThroughStream {
 			onFinished: this.#onFinished,
 			unpipeEvent: this.#unpipeEvent,
 		});
+		this.#streamPromises.set(stream, streamPromise);
 
 		stream.pipe(this, {end: false});
 	}
 
-	remove(stream) {
+	async remove(stream) {
 		validateStream(stream);
 
 		if (!this.#streams.has(stream)) {
 			return false;
 		}
 
+		const streamPromise = this.#streamPromises.get(stream);
+		if (streamPromise === undefined) {
+			return false;
+		}
+
+		this.#streamPromises.delete(stream);
+
 		stream.unpipe(this);
```

Comment:
`stream.unpipe()` emits the `unpipe` event. We listen to it using `once(stream, 'unpipe`)`. There are a few promises involved, which might this takes a few ticks to resolve.

Due to using events, we cannot simply `await stream.unpipe(this)` though. So we need to keep track of ongoing promises in a `WeakMap` instead. Those promises never reject, so cannot create unhandled promise rejections.